pkg/cluster/space/openebs.go

```diff
@@ -0,0 +1,599 @@
+package clusterspace
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"code.cloudfoundry.org/bytefmt"
+	"gopkg.in/yaml.v2"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
+
+	"github.com/google/uuid"
+	"github.com/replicatedhq/kurl/pkg/k8sutil"
+)
+
+// OpenEBSChecker checks if we have enough disk space on the cluster to migrate volumes to openebs.
+type OpenEBSChecker struct {
+	deletePVTimeout time.Duration
+	kcli            kubernetes.Interface
+	log             *log.Logger
+	image           string
+	srcSC           string
+	dstSC           string
+}
+
+// parseDFContainerOutput parses the output (log) of the 'disk available' pod. the output of the
+// container is expected to be the default df command output (with bytes as unit or measurement):
+//
+// Filesystem     1K-blocks     Used Available Use% Mounted on
+// /dev/sda2       61608748 48707392   9739400  84% /data
+//
+// the openebs node volume is mounted under /data inside the pod. this function returns the
+// amount of used and available space as bytes.
+func (o *OpenEBSChecker) parseDFContainerOutput(output []byte) (int64, int64, error) {
+	buf := bytes.NewBuffer(output)
+	scanner := bufio.NewScanner(buf)
+	for scanner.Scan() {
+		words := strings.Fields(scanner.Text())
+		if len(words) == 0 {
+			continue
+		}
+
+		// lastpos is where the mount point lives.
+		lastpos := len(words) - 1
+		if words[lastpos] != "/data" || len(words) < 5 {
+			continue
+		}
+
+		// pos is the position where the actual available space is.
+		pos := len(words) - 3
+		freeBytes, err := strconv.ParseInt(words[pos], 10, 64)
+		if err != nil {
+			return 0, 0, fmt.Errorf("failed to parse %q as available space: %w", words[pos], err)
+		}
+
+		// pos is now the position where the actual used space is.
+		pos = len(words) - 4
+		usedBytes, err := strconv.ParseInt(words[pos], 10, 64)
+		if err != nil {
+			return 0, 0, fmt.Errorf("failed to parse %q as used space: %w", words[pos], err)
+		}
+
+		return freeBytes, usedBytes, nil
```

Comment:
I would REALLY prefer a regex-based implementation here (with capture groups)

Regexes may be brittle and inscrutable, but so's using word order directly

Reply:
On the other hand it's pretty common to parse the output of linux command line with awk